---
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
title: "SeqPlots R workflow"
author: "Przemyslaw Setmpor"
date: "`r format(Sys.Date(), '%d %B %Y')`"
package: "`r paste('SeqPlots', packageVersion('seqplots'))`"
abstract: >
  SeqPlots is a tool for plotting average track signals (e.g.
  read coverage) and sequence motif densities over user specified genomic
  features. The data can be visualized in linear plots with error
  estimates or as series of heatmaps that can be sorted and clustered. The
  software can be run locally on a desktop or deployed on a server and
  allows easy data sharing. SeqPlots pre-calculates and stores binary
  result matrices, allowing rapid plot generation. Plots can also be run
  in batch.
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---


```{r echo=FALSE, results='asis', echo=FALSE}
# knit the first three lines of first.Rmd
cat(knitr::knit_child(
    text = gsub(
        "img/",
        "../inst/seqplots/www/help/img/",
        readLines('../inst/seqplots/www/help/QuickStart.md')
    ),
    quiet = TRUE
), sep = '\n')
```


Session Information
===================
=======
title: "Quick start guide"
author: "Przemyslaw Stempor"
date: "September 12, 2014"
output:
  BiocStyle::html_document
---

<!-- 
%\VignetteIndexEntry{SeqPlots GUI QuickStart}
%\VignetteEngine{knitr::knitr}
-->


```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```


Quick start guide
=================

1. Start the SeqPlots GUI with `seqplots::run()` command in R. After successful initiation the web interface should automatically open in your default web browser. If you are web server user just navigate your browser to server address.

    ![The SeqPlots interface in web browser](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_01.png)

2. Upload some feature (BED or GFF) and track (BigWig or WIG) files. They can be gzip compressed (e.g. file1.bed.gz). Press green "Add files..." button or just drag and drop files into the window. The ready to upload files will show up in upload window, where you select user name, reference genome and optionally add some comments.

    ![File upload panel](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_02.png)

3. When all is done press blue "Start upload" button. After upload and processing is done the green "SUCCESS" label should show. It means that file is on the registered and ready to use. Occasionally the file might be mot formatted properly or chromosome names might not agree with reference genome. In such case a verbose error will window appears and file as labeled as "ERROR". For further information please refer to [**errors chapter**](SeqPlotsGUI.html#Errors).
    
    ![File upload progress infoermation](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_03.png) 


4. Dismiss upload window and press blue "New plot set" button on side panel. This will bring up file management window. In file management window select at least one file from "Features" tab and either one or more file from "Tracks" file or sequence motif(s). The sequence motifs and tracks can be processed and plotted together.
    
    ![File management panel](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_04.png) 


5. When you decided which files/motifs to plot it is time to set up the processing options. You can find these in the button of plotting window. For first plot you should do just fine with default options, to learn more check [**this**](SeqPlotsGUI.html) section.
    
    ![Plot set calculation options](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_05.png)


6. After options are set up press blue "Run calculation" button. This will dismiss the file management window and show processing message. Here you can observe the progress of the task and optionally cancel it if no longer required or you forgot to add some very important file to the plot-set (cancel available only in multi process mode).
    
    ![Plot set calculation progress window](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_06.png)


7. After some time the calculation will finish (fingers crossed, without the error) and you will be able to see plot set array. In here you can choose which feature-track or feature-motif pairs to plot. Choose one or more checkboxes and press grey "Line plot" button (or hit RETURN from your keyboard). You can also check "Reactive plotting" checkbook in the bottom of "Batch operations and setup" tab - it will automatically apply changes to plots as soon as you changed any plot-related option.
    
    ![The plot selection grid](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_07.png)


8. Congratulation! Your First plot is complete, you can see the preview of it on the side panel.
    
    ![The plot preview panel](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_08.png)


9. You are able to set up labels, titles, font sizes, legends and many more on side panel tabs, take a look at [**this**](SeqPlotsGUI.html#Plotting) chapter.
    
    ![Plot settings tabs](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_09.png)

10. By clicking the plot preview you can enlarge it for better view. When everything is ready you can get the plot as PDF by clicking green "Line plot" button just on the top of side panel.

    ![Average plot example](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_10.png)


11. You can also visualize the signal as a heatmap. Please note that heatmap plotting is possible only for single feature file or files containing exactly the same number of genomic ranges (which will become the rows of heat map). For heatmap you can choose to sort and/or cluster it using k-means. To learn more about heatmaps click [**here**](SeqPlotsGUI.html).
    
    ![Heatmap settings tab](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_11.png)

12. Similarly to line plot you can grab the heatmap PDF using 'Heatmap' button just on the top of side panel.
Just attached to it, there is small button allowing you to get cluster definitions.

    ![Heatmap example](../inst/seqplots/www/help/img/01_QuickStart/01_QuickStart_12.png)


Session Information
-------------------
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
```{r echo=FALSE}
sessionInfo()
```
